mechanicalsoup/stateful_browser.py

```diff
@@ -230,16 +230,32 @@ def links(self, url_regex=None, link_text=None, *args, **kwargs):
                          if a.text == link_text]
         return all_links
 
-    def find_link(self, *args, **kwargs):
+    def find_link(self, link=None, *args, **kwargs):
```

Comment:
Won't this break backwards compatibility for anyone who has specified `*args`? Unfortunately this whole situation is muddled by the `*args` situation, which doesn't seem to have a satisfying resolution. :/

Actually, in this one case, I think we can at least not make the situation worse by keeping
```python
if not hasattr(link, 'attrs') or 'href' not in link.attrs:
    link = self.find_link(*args, **kwargs)
```
in both `follow_link` and `download_link` instead of
```python
link = self.find_link(link, *args, **kwargs)
```
It's only 1 line of code duplicated, but I think avoids a lot of complexity in `find_link`.

Reply:
(And, if it wasn't clear, we would not need the `link` param in `find_link` in this case.)

Reply:
> Won't this break backwards compatibility for anyone who has specified *args?

Not really. `*args` is for positional arguments, so what I did is to special-case the first positional argument. A call to `find_link(foo, bar, boz)` was earlier setting `args=(foo, bar, boz)` and now `link=foo, args=(bar, boz)`. But since I later forward `(link,) + args`, the result is the same.

I can think of one case where this was not backward-compatible: `find_link(None, foo)`, i.e. pass `None` as a positional argument and expect it to be forwarded as `url_regex` (so that `foo` becomes the `link_text` argument). I'll fix it in my next push.

Reply:
I really think any change to the `find_link` interface is overcomplicating this...

Reply:
I actually think that any difference between `find_link` and `list_link`'s list of argument is a source of problem. Changing the first to match the second will fix #178 as a side effect.